### ISSUE TITLE
🐛 [Bug Report] Stair of Desteny

### ISSUE THREAD

**Issue by miraco** [cmangos/issues#2061]
## 🐛 Bugreport
After finishing the last RP event for friendly npcs (atleast i hope) at stair of desteny, i wanna start rescripting the Stair of Desteny invaders event. I will just take this bugreport for research.

### Expected behavior
Short Version:
- 3 Spawn Points
- 4 variants of Groups.
- getting spawned, even if no pit lord is a live

Pit Lord:
Getting spawned between the 2 towers.
![spawn pit](https://user-images.githubusercontent.com/6261245/68291148-8a5b9a00-0089-11ea-8514-cd09b8a2a63e.jpg)
![pitlord](https://user-images.githubusercontent.com/6261245/68291182-9cd5d380-0089-11ea-823d-34f65af9b3af.jpg)
Moving to the dark portal. When waypoints finished he will start doing emote (random timer) and spawn infernals/ rain of fire.

AI
`Casting Teleport Visual Only 41232`

First Group:
8 Fel Soldiers getting spawned in the gateways. 
![fel soldier left](https://user-images.githubusercontent.com/6261245/68291548-62b90180-008a-11ea-9aa6-9872da335e84.jpg)
![fel soldier right](https://user-images.githubusercontent.com/6261245/68291549-62b90180-008a-11ea-8be0-523bd96a0f0c.jpg)

Running to the path and getting 1 big Group.
![fel soldier group](https://user-images.githubusercontent.com/6261245/68291581-73697780-008a-11ea-9b48-42996cc6d81a.jpg)

Moving to the dark portal. When they reach the stair, they start attaking the defenders.  If no defenders can stop them, they walk into the dark portal, and disappear after 2 seconds.

![soldier end](https://user-images.githubusercontent.com/6261245/68291765-d5c27800-008a-11ea-8bd4-dc7fd28cda53.jpg)

AI:
`Cast Teleport Visual Only on Spawn`

Second Group
8 Fel Soldiers 2 Wrath Masters spawned in the gateways.

Third Group
1 Wrath Master 4 Fel Soldiers, spawned between the 2 towers just like the pit lord.
![big group](https://user-images.githubusercontent.com/6261245/68292059-6436f980-008b-11ea-9372-ea8800e2d7f5.jpg)

Fourth Group
This is a strange group and i only saw her once for noce. 3 Wrath Master, spawned at the 2 towers.
![wrath group](https://user-images.githubusercontent.com/6261245/68292044-5e411880-008b-11ea-997f-4ce0b303c629.jpg)

The biggest challange will to set them as groups. 

I will post updates when i figured out more.
Client Version: 
2.4.3

**Comment by killerwife** [comment]
I suggest that if you wish to do this, do it in C++. The scale is huge. Currently there is a large part done in DB. @MantisLord if you wish to explain more.
If it's done, my bad.

**Comment by miraco** [comment]
I guess some things you can leave in DB like the Pit lord. He is only doing the waypoints and then doing some emotes/spawningn infernals.
The rest, group handling and spawning waves ill try to do via c++ but i will first have to do some more research.

Its not fully working as far as i saw on my local server.

**Comment by AnonXS** [comment]
Core: Quest_10985_A_Distraction_For_Akama (a3f17570739566551b15f75f6a363663dee87239) - by Sneaky might be a good inspiration when doing this Event in Core, might be worth looking up when it's exported.

**Comment by killerwife** [comment]
It already is in cmangos ^^

**Comment by AnonXS** [comment]
You're right! https://github.com/cmangos/mangos-tbc/commit/0c358c0887eaefac763aba56b79122fb7a89a94d

**Comment by MantisLord** [comment]
The full event was scripted in the DB here.
https://github.com/cmangos/tbc-db/commit/68ea38c08ea8bb0bdca23f522142ac73113b297d

The only thing that's lacking as far as I'm concerned is some of the formations movement which really is only noticeable when the packs reach the "staging area" in front of Pit Lord.

As far as number of packs, the timing between them, etc... Most of that should be pretty spot on already. I analyzed retail the same way you did to create them.

When you say it wasn't working for you locally, some more detail would be nice. What specifically wasn't working?

**Comment by MantisLord** [comment]
Also I might add that the way the mobs move in this staging area would not necessarily be solved by a simple follow angle adjustment. They all spread out at once in a straight line before moving forward - it's sorta unlike most other groups in the world I've seen before.

**Comment by miraco** [comment]
The first bug i saw, were more npcs in the world (not moving?) like this group.
![WoWScrnShot_110619_175827](https://user-images.githubusercontent.com/6261245/68319753-1805ac80-00bf-11ea-9397-88bf4bf8a8d7.jpg)

We will have to delete them, npcs with entry 18944 (Fel Solider) and 19005 (Wrath Master) are only used for the event with waypoints not just staying in the world.

Afaik you currently handled it with just spawning them and let them move based on her GUID? So every creature of the group has a waypoint.

So its always just 2/3 waves that are up. On retail i saw like 4/5 or even more.

The thing happening in the staging area looks realy wierd tbh. Not sure we are able to figure out how blizz is handeling this.
My opinion on this were to create a base line with 10-15 fixed points. Searching for a free spot and running on it. When no spot is free send some of the creatures that are already on a spot to the stair and start fighting. (Could handle this via pit lord or invisible triggers) so we get free spots before next wave spawns/reach waypoints)

**Comment by MantisLord** [comment]
Something is off. There should be no static mobs like that, they all have movement.

I'll look into this tonight and report back here.

**Comment by MantisLord** [comment]
@miraco Thank you for bringing this to my attention. The data was not in a state that I expected.
I went ahead and removed the static spawns that you took screencaps of above.

As for increasing the number of waves that spawn at once, or adjusting the spawntimes of each wave, this is an area which could do with some improvement, but I still think this event as a whole is appropriate for scripting in the DB.

**Comment by MantisLord** [comment]
Also here are the notes I took about the waves when I first reviewed this event on retail a couple years ago (pulled from SQL update file linked above).

WAVE 1: 4 FEL SOLDIER - TELEPORT IN FROM GATEWAY SHAADRAZ
1 min respawn

WAVE 2: 4 FEL SOLDIER - TELEPORT IN FROM GATEWAY MURKETH
1 min respawn

WAVE 3: 1 WRATH MASTER + 4 FEL SOLDIER - TELEPORT IN BETWEEN LARGE PILLARS
3 min respawn

WAVE 4: 1 WRATH MASTER + 4 FEL SOLDIER - TELEPORT IN BETWEEN LARGE PILLARS
3 min respawn

WAVE 5: 1 WRATH MASTER + 4 FEL SOLDIER - TELEPORT IN FROM GATEWAY MURKETH
ONLY SPAWNS WHEN THE PIT COMMANDER IS PASSING BY ON HIS WAY TO THE STAIR OF DESTINY
1 hour respawn and triggered respawn by Pit Commander

WAVE 6: 1 WRATH MASTER + 4 FEL SOLDIER - TELEPORT IN FROM GATEWAY SHAADRAZ
ONLY SPAWNS WHEN THE PIT COMMANDER IS PASSING BY ON HIS WAY TO THE STAIR OF DESTINY
1 hour respawn and triggered respawn by Pit Commander

Your discovery of the 3 Wrath Master group seems to be something we're lacking currently.

**Comment by killerwife** [comment]
@MantisLord close this if you can. We now do have formations :P
